### PR TITLE
Refac/layering

### PR DESCRIPTION
**Pull request by ramon-reichert**
The packages are separated, but the responsibilities not yet. If you can we could talk tomorrow to finish it together.

### CODE REVIEW

**Comment by ramon-reichert** [comment]
I've written doubts on comments at main.go and handlers.go.

File: cmd/api/book/service.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
+package book
+
+import (
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type ServiceAPI interface {
+	ArchiveStatusBook(id uuid.UUID) (Book, error)
```
Comment (cesarFuhr) [inline cmd/api/book/service.go]:
```suggestion
	ArchiveBook(id uuid.UUID) (Book, error)

File: cmd/api/book/service.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	CreateBook(bookEntry Book) (Book, error)
+	GetBook(id uuid.UUID) (Book, error)
+	UpdateBook(bookEntry Book, id uuid.UUID) (Book, error)
+}
+
+type Repository interface {
+	ArchiveStatusBook(id uuid.UUID, archived bool) (Book, error)
```
Comment (cesarFuhr) [inline cmd/api/book/service.go]:
```suggestion
	SetBookArchiveStatus(id uuid.UUID, archived bool) (Book, error)

File: cmd/api/http/handlers.go [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+
+type BookHandler struct {
+	bookService book.ServiceAPI
+}
+
+func NewBookHandler(bookService book.ServiceAPI) *BookHandler {
+	return &BookHandler{bookService: bookService}
+}
+
+//==========HTTP ADDRESSERS:===========
```
Comment (cesarFuhr) [inline cmd/api/http/handlers.go]: I don't think this name is accurate. Maybe router or mux.
I actually would start to question the need of these section comments, are they needed now that our code is spread in more files?

Comment (ramon-reichert) [reply cmd/api/http/handlers.go]: I can take it of.

File: cmd/api/book/service.go (hunk 2 above)
Comment (ramon-reichert) [reply cmd/api/book/service.go]: Ok. I've got it. Changing...

**Comment by ramon-reichert** [comment]
We left behind the moving of json tags to http package. I'll take a look at this.

File: cmd/api/book/book.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+		}
+	}
+
+	pagesTotal = int(math.Ceil(float64(itemsTotal) / float64(pageSize)))
+	if page > pagesTotal {
+		return 0, 0, 0, ErrResponseQueryPageOutOfRange
+	}
+
+	return pagesTotal, page, pageSize, nil
+}
```
Comment (cesarFuhr) [inline cmd/api/book/book.go]: This is not related to the book entity at all, I don't see any mention of a book in these pagination/total parameters functions.
Please move this to the handlers files.

File: Makefile
Comment (cesarFuhr) [inline Makefile:1]: Please remove the formatting changes, they aren't adding anything.

File: cmd/api/book/service.go [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+type PagedBooks struct {
+	PageCurrent int    `json:"page_current"`
+	PageTotal   int    `json:"page_total"`
+	PageSize    int    `json:"page_size"`
+	ItemsTotal  int    `json:"items_total"`
+	Results     []Book `json:"results"`
+}
+
+func (s *Service) ListBooks(query url.Values) (PagedBooks, error) {
```
Comment (cesarFuhr) [inline cmd/api/book/service.go]: You shouldn't be dealing with `url.Values` here. This is specific to the HTTP layer.
What you can do to pass the search parameters to the `ListBooks` method is create a `request` type.
```go
type ListBooksRequest struct {
    Name string
    MinPrice float
//...
}

Comment (cesarFuhr) [reply cmd/api/book/service.go]: Then you would extract the values from the `http.Request` in the HTTP layer.

File: cmd/api/database/database.go [hunk 6]
```diff
@@ -1,76 +1,90 @@
...
+			return book.Book{}, fmt.Errorf("archiving on db: %w", err)
 		}
 	}
 
 	return bookToReturn, nil
 }
 
-func countRows(name string, minPrice32, maxPrice32 float32, archived bool) (int, error) {
+/* Counts how many rows in db fit the specified filter parameters. */
+func (store *Store) CountRows(name string, minPrice32, maxPrice32 float32, archived bool) (int, error) {
```
Comment (cesarFuhr) [inline cmd/api/database/database.go]: This method needs to be closer to the `ListBooks` method.
I think a better name would be `CountListBooksTotals` or simply `ListBooksTotals`.

File: cmd/api/database/database.go [hunk 7]
```diff
@@ -92,8 +106,42 @@ func countRows(name string, minPrice32, maxPrice32 float32, archived bool) (int,
...
+	if err != nil {
+		return book.Book{}, fmt.Errorf("storing on db: %w", err)
+	}
+
+	return bookToReturn, nil
+}
+
+/* Searches a book in database based on ID and returns it if succeed. */
+func (store *Store) GetBookByID(id uuid.UUID) (book.Book, error) {
+	sqlStatement := `SELECT id, name, price, inventory, created_at, updated_at, archived FROM bookstable WHERE id=$1;`
```
Comment (cesarFuhr) [inline cmd/api/database/database.go]: Please format this `sql` query like you did with the others. They read way better.

File: cmd/api/main.go [hunk 8]
```diff
@@ -1,433 +1,57 @@
...
 	}
-
-	return sortBy, sortDirection, true
 }
 
-func main() {
+func run() error {
 	//connect to db:
-	dbObject, err := connectDb()
+	connStr := os.Getenv("DATABASE_URL") //I'M STILL CONFUSED ABOUT THAT. IN DATABASE_TEST WE CHANGED THE ENV BY A LITERAL
```
Comment (cesarFuhr) [inline cmd/api/main.go]: We can change the test code to read from the environment too, would be a nice improvement.

**Review by cesarFuhr — CHANGES_REQUESTED** [review]

File: Makefile
Comment (ramon-reichert) [reply Makefile:1]: Ok. I did not even see the changes. It is just blank spaces added accidentally.

File: cmd/api/main.go (hunk 8 above)
Comment (ramon-reichert) [reply cmd/api/main.go]: I took a look at the old commits. now, to remember how we used to deal with the connStr in database_test before.
I've got it:
We used to SET (not GET!!) the connStr as an ENV because ConnectDB() did not receive anything:
`func TestMain(m *testing.M) {
	os.Setenv("DATABASE_URL", "postgres://root:root@localhost:5432/booksdb?sslmode=disable")
	db, err := ConnectDb()`

Now that `database.ConnectDb(connStr)` receive the connStr, we do not need the ENV anymore.

File: cmd/api/database/database.go (hunk 7 above)
Comment (ramon-reichert) [reply cmd/api/database/database.go]: ok, done.

File: cmd/api/database/database.go (hunk 6 above)
Comment (ramon-reichert) [reply cmd/api/database/database.go]: Done.

File: Makefile
Comment (ramon-reichert) [reply Makefile:1]: Ok. I did not even see the changes. It is just blank spaces added accidentally.

File: cmd/api/book/service.go (hunk 5 above)
Comment (ramon-reichert) [reply cmd/api/book/service.go]: I've tried to pass a variable of type http.ListBooksRequest to service.go, but then it felt in a cycle import error. So I had to pass the list of parameters extensively. It worked, at least.

File: cmd/api/book/book.go (hunk 4 above)
Comment (ramon-reichert) [reply cmd/api/book/book.go]: Ok.

File: cmd/api/book/service.go (hunk 5 above)
Comment (ramon-reichert) [reply cmd/api/book/service.go]: In the middle of the night I realized that I could create the struct type at service.go side!

File: cmd/api/main.go (hunk 8 above)
Comment (cesarFuhr) [reply cmd/api/main.go]: Yeah, but its better to control the connection string from the outside.
We can have a quick pairing session for me to show you why is that.

File: Makefile [hunk 9]
```diff
@@ -12,3 +12,6 @@ test-env-up:
 
 test-env-down:
 	docker compose -f docker-compose.test.yml down
+
+
+
```
Comment (cesarFuhr) [inline Makefile]: Remove this ones too.

File: cmd/api/book/book.go [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+type Book struct { //IS IT POSSIBLE TO MOVE JSON TAGS TO HTTP PACKAGE??
+	ID        uuid.UUID `json:"id"`
+	Name      string    `json:"name"`
+	Price     *float32  `json:"price"`
+	Inventory *int      `json:"inventory"`
+	CreatedAt time.Time `json:"-"`
+	UpdatedAt time.Time `json:"-"`
+	Archived  bool      `json:"archived"`
+}
```
Comment (cesarFuhr) [inline cmd/api/book/book.go:19]: We should, but I want to finish this refactor first, we can deal with that later. I will make sure we have a ticket to address it.

File: cmd/api/http/handlers.go [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+		archived = true
+	}
+
+	page, pageSize, valid := extractPageParams(query)
+	if !valid {
+		responseJSON(w, http.StatusBadRequest, book.ErrResponseQueryPageInvalid)
+		return
+	}
+
+	pagedBooks, err := h.bookService.ListBooks(name, minPrice32, maxPrice32, sortBy, sortDirection, archived, page, pageSize)
```
Comment (cesarFuhr) [inline cmd/api/http/handlers.go]: Define a request type as we discussed.
It should be defined inside the `book` package, that will avoid the cyclic dependency issue you had.

**Review by cesarFuhr — CHANGES_REQUESTED** [review]
Almost there.

File: cmd/api/book/book.go (hunk 10 above)
Comment (cesarFuhr) [reply cmd/api/book/book.go:19]: https://trello.com/c/scwnaydr/56-refactor-define-request-types

File: cmd/api/book/book.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+type ListBooksRequest struct {
+	Name          string
+	MinPrice      float32
+	MaxPrice      float32
+	SortBy        string
+	SortDirection string
+	Archived      bool
+	Page          int
+	PageSize      int
+}
```
Comment (cesarFuhr) [inline cmd/api/book/book.go]: Please move this definition closer to the method that uses it.

**Review by cesarFuhr — CHANGES_REQUESTED** [review]
One last change.

**Review by cesarFuhr — APPROVED** [review]
